File: samples/java/DeliveryDispatch/Server/CustomerGateway/src/main/java/systems/zlink/samples/deliverydispatch/server/customergateway/spots/handlers/SubscribeDeliveryActorHandler.java
package systems.zlink.samples.deliverydispatch.server.customergateway.spots.handlers;

import systems.zlink.framework.ZLinkMessageContext;
import systems.zlink.framework.spots.ZLinkEntrySpotActorRequestHandler;
import systems.zlink.samples.deliverydispatch.server.customergateway.CustomerActor;
import systems.zlink.samples.deliverydispatch.server.customergateway.spots.CustomerEntrySpot;
import systems.zlink.samples.deliverydispatch.shared.contracts.Messages;

import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;

public final class SubscribeDeliveryActorHandler
        implements ZLinkEntrySpotActorRequestHandler<
                CustomerEntrySpot,
                CustomerActor,
                Messages.SubscribeDeliveryReq,
                Messages.SubscribeDeliveryRes> {
    @Override
    public CompletionStage<Messages.SubscribeDeliveryRes> handle(
            CustomerEntrySpot entrySpot,
            CustomerActor actor,
            ZLinkMessageContext context,
            Messages.SubscribeDeliveryReq request) {
        return CompletableFuture.completedFuture(entrySpot.subscribe(actor, request));
    }
}

File: samples/kotlin/DeliveryDispatch/Server/CustomerGateway/src/main/kotlin/systems/zlink/samples/kotlin/deliverydispatch/server/customergateway/spots/handlers/SubscribeDeliveryActorHandler.kt
package systems.zlink.samples.kotlin.deliverydispatch.server.customergateway.spots.handlers

import systems.zlink.framework.ZLinkMessageContext
import systems.zlink.framework.kotlin.ZLinkSuspendingEntrySpotActorRequestHandler
import systems.zlink.samples.kotlin.deliverydispatch.server.customergateway.CustomerActor
import systems.zlink.samples.kotlin.deliverydispatch.server.customergateway.spots.CustomerEntrySpot
import systems.zlink.samples.kotlin.deliverydispatch.shared.contracts.SubscribeDeliveryReq
import systems.zlink.samples.kotlin.deliverydispatch.shared.contracts.SubscribeDeliveryRes

class SubscribeDeliveryActorHandler :
    ZLinkSuspendingEntrySpotActorRequestHandler<
        CustomerEntrySpot,
        CustomerActor,
        SubscribeDeliveryReq,
        SubscribeDeliveryRes,
    > {
    override suspend fun handle(
        entrySpot: CustomerEntrySpot,
        actor: CustomerActor,
        context: ZLinkMessageContext,
        request: SubscribeDeliveryReq,
    ): SubscribeDeliveryRes = entrySpot.subscribe(actor, request)
}
